src/toolbox/xor.rs

```diff
@@ -0,0 +1,270 @@
+use curve25519_dalek::scalar::Scalar;
+use rand::{CryptoRng, RngCore};
+use rand::prelude::ThreadRng;
+use crate::toolbox::secrets;
+use log::{trace, debug, info};
+
+/// A SecretSharing implementation that is based on the all-shares XOR method.  In particular, you must have every outstanding share
+/// of a secret in order to reconstruct that secret.  Because it is the simplest method, it has no interesting parameters.
+pub struct Xor<R: CryptoRng + RngCore> {
+    rng: R,
+}
+
+impl<R> Xor<R> where R: RngCore + CryptoRng {
+    /// Create a new Xor object with the provided RNG implementation.
+    pub fn new(rng: R) -> Self {
+        Xor {
+            rng
+        }
+    }
+}
+
+impl Xor<ThreadRng> {
+    /// Create a new Xor object with a default RNG.
+    pub fn new_without_rng() -> Self {
+        Xor { rng: rand::thread_rng() }
+    }
+}
+
+impl<R> secrets::SecretSharing for Xor<R> where R: RngCore + CryptoRng {
+    fn share(&mut self, secret: &Scalar, nr_of_shares: usize) -> Result<Vec<Scalar>, String> {
+        info!("Sharing {} pieces of {:?}", nr_of_shares, secret);
+        let mut shares: Vec<Scalar> = Vec::new();
+        
+        // generate n - 1 random shares; the zero-th share is the secret
+        for _ in 1..nr_of_shares {
+            shares.push(Scalar::random(&mut self.rng));
+        }
+
+        // calculate the final share as a function of the random shares and the secret
+        match xor_many_scalars(secret, &shares) {
+            Ok(val) => shares.push(val),
+            Err(e) => return Err(e),
+        }
+
+        return Ok(shares);
+    }
+
+    /// For Xor, sparse_shares should include a None value for each share you want us to generate
+    fn complete(&mut self, secret: &Scalar, sparse_shares: &Vec<Option<Scalar>>) -> Result<Vec<Scalar>, String> {
+        let mut empties = 0;
+        let mut shares = Vec::new();
+        let mut new_shares = Vec::new();
+        for share in sparse_shares.iter() {
+            if share.is_some() {
+                shares.push(share.unwrap());
+            } else {
+                if empties != 0 {     // the "zero-th" empty will be the one we have to XOR-calculate
+                    trace!("Pushing a random Scalar to extra empty");
+                    new_shares.push(Scalar::random(&mut self.rng));
```

Comment:
Is this a state that we want to handle or shouldn't this just output an error?

Reply:
In this loop, `sparse_shares` is an n+m-sized Vector.  There are n provided shares (these are generated in `prover.commit()` for each unknown key) and m empty shares (these are the ones we're supposed to `complete`).  For the XOR algorithm specifically, we randomize m-1 shares, and then calculate the final share by XOR'ing all the previous shares along with the secret.  That final piece is part of what this loop does.

Reply:
But this means that not all other shares were given, aren't we assuming we receive all other shares that we have to XOR, without creating any new random shares here?

Reply:
That's true in `reconstruct()`, which starts around line 90, but this is within `complete()`, which is specifically tasked with creating additional shares.